olp-cpp-sdk-dataservice-read/tests/CatalogRepositoryTest.cpp

```diff
@@ -105,126 +111,245 @@ class CatalogRepositoryTest : public ::testing::Test {
+  const auto request = read::CatalogVersionRequest()
+                           .WithFetchOption(olp::dataservice::read::CacheOnly)
+                           .WithStartVersion(-1);
+
+  auto cached_version = read::model::VersionResponse();
+  cached_version.SetVersion(1);
+  EXPECT_CALL(*cache_, Get(kLatestVersionCacheKey, _))
+      .WillOnce(testing::Return(cached_version));
+
+  ApiLookupClient lookup_client(kHrn, settings_);
+  repository::CatalogRepository repository(kHrn, settings_, lookup_client);
+
+  std::promise<read::CatalogVersionResponse> promise;
+  auto future = promise.get_future();
+  repository.GetLatestVersion(request,
+                              [&](read::CatalogVersionResponse response) {
+                                promise.set_value(std::move(response));
+                              });
+
+  ASSERT_EQ(future.wait_for(kMaxWaitMs), std::future_status::ready);
+  const auto result = future.get();
+
+  EXPECT_TRUE(result);
+  EXPECT_EQ(result.GetResult().GetVersion(), cached_version.GetVersion());
+}
+
+TEST_F(CatalogRepositoryTest, AsyncGetLatestVersionOnlineOnlyNotFound) {
+  auto request =
+      read::CatalogVersionRequest().WithFetchOption(read::OnlineOnly);
+
+  ON_CALL(*cache_, Get(_, _)).WillByDefault(testing::WithoutArgs([]() {
+    ADD_FAILURE() << "Cache should not be used in OnlineOnly request";
+    return boost::any{};
+  }));
+
   EXPECT_CALL(*cache_, Get(testing::Eq(kMetadataCacheKey), _))
       .WillOnce(testing::Return(boost::any()));
 
   EXPECT_CALL(*network_, Send(IsGetRequest(kLookupMetadata), _, _, _, _))
-      .Times(1)
-      .WillOnce(ReturnHttpResponse(olp::http::NetworkResponse().WithStatus(
-                                       olp::http::HttpStatusCode::NOT_FOUND),
-                                   ""));
+      .WillOnce(ReturnHttpResponse(
+          http::NetworkResponse().WithStatus(http::HttpStatusCode::NOT_FOUND),
+          ""));
 
   ApiLookupClient lookup_client(kHrn, settings_);
   repository::CatalogRepository repository(kHrn, settings_, lookup_client);
-  auto response = repository.GetLatestVersion(request, context);
 
-  EXPECT_FALSE(response.IsSuccessful());
+  std::promise<read::CatalogVersionResponse> promise;
+  auto future = promise.get_future();
+  const auto response = repository.GetLatestVersion(
+      request, [&](read::CatalogVersionResponse response) {
+        promise.set_value(std::move(response));
+      });
+
+  ASSERT_EQ(future.wait_for(kMaxWaitMs), std::future_status::ready);
+  const auto result = future.get();
+
+  EXPECT_FALSE(result);
+  EXPECT_EQ(result.GetError().GetErrorCode(), client::ErrorCode::NotFound);
 }
 
-TEST_F(CatalogRepositoryTest, GetLatestVersionOnlineOnlyFound) {
-  olp::client::CancellationContext context;
+TEST_F(CatalogRepositoryTest, AsyncGetLatestVersionOnlineOnlyForbidden) {
+  const auto request =
+      read::CatalogVersionRequest().WithFetchOption(read::OnlineIfNotFound);
```

Comment:
You have `OnlineIfNotFound` request, but the name of the test contains `OnlineOnly`. Also, line 316 contains a mention of `OnlineOnly`.